### PR TITLE
ADEN-10302 BigFancyAd* servable from SafeFrame

### PR DESCRIPTION
**Pull request by fraszczakszymon**

### CODE REVIEW

File: src/ad-products/templates/uap/universal-ad-package.ts [hunk 1]
```diff
@@ -72,7 +73,7 @@ export interface UapParams {
 	lineItemId: string;
 	loadMedrecFromBTF: boolean;
 	moatTracking: boolean;
-	player: HTMLElement;
+	player: string;
```
Comment (Bielik20) [inline src/ad-products/templates/uap/universal-ad-package.ts:76]: How does it work now?

File: src/ad-engine/utils/iframe-builder.ts [hunk 2]
```diff
@@ -1,10 +1,16 @@
-import { AdSlot } from '../models';
-
 export class IframeBuilder {
-	create(adSlot: AdSlot): HTMLIFrameElement {
-		const doc = adSlot.getElement();
+	create(element: HTMLElement, htmlContent: string = null): HTMLIFrameElement {
```
Comment (Bielik20) [inline src/ad-engine/utils/iframe-builder.ts:2]: Why default to null?
```suggestion
	create(element: HTMLElement, htmlContent?: string): HTMLIFrameElement {
```

File: src/ad-products/templates/uap/universal-ad-package.ts (hunk 1 above)
Comment (fraszczakszymon) [reply src/ad-products/templates/uap/universal-ad-package.ts:76]: I believe it was a typo. `player` was a string from the very beginning.

File: src/ad-products/templates/safe-uap/safe-fan-takeover-config-loader.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			clickThroughUrl: 'https://fandom.com/',
+			thumbnail:
+				'//static.nocookie.net/fandom-ae-assets/programmatic/latest/campaign/1/thumbnail.jpg',
+			vast: '//static.nocookie.net/fandom-ae-assets/programmatic/latest/campaign/1/vast.xml',
+			campaignId,
+		};
+	}
+}
+
+export const safeFanTakeoverConfigLoader = new SafeFanTakeoverConfigLoader();
```
[the remark this answers is not in the document]
Comment (fraszczakszymon) [reply src/ad-products/templates/safe-uap/safe-fan-takeover-config-loader.ts:65]: 👍

File: src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+		`;
+	}
+
+	private setupElements(iframeDocument: Document): void {
+		this.adContainer = iframeDocument.getElementById('adContainer');
+		this.defaultBackground = iframeDocument.getElementById('defaultBackground');
+		this.resolvedBackground = iframeDocument.getElementById('resolvedBackground');
+
+		[this.defaultBackground, this.resolvedBackground].forEach((element) => {
+			element.addEventListener('click', () => top.open(this.config.clickThroughUrl, '_blank'));
```
[the remark this answers is not in the document]
Comment (fraszczakszymon) [reply src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts:55]: Discussed on slack. `top` is the shortest one ;)

File: src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+			});
+			thumbnailImage.src = this.config.thumbnail;
+		}
+	}
+
+	private initTemplate(): void {
+		const deviceType = context.get('state.isMobile') ? 'mobile' : 'desktop';
+		const state = this.getStateSetup(deviceType);
+		const platformConfig = this.config[deviceType];
+		const templateName = universalAdPackage.isFanTakeoverLoaded() ? 'bfab' : 'bfaa';
```
[the remark this answers is not in the document]
Comment (fraszczakszymon) [reply src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts:80]: Discussed on slack. We've agreed to make a decision based on this logic, for now.

File: src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+				video: {
+					thumb: this.config.thumbnail,
+				},
+				state,
+			},
+
+			slotName: this.adSlot.getSlotName(),
+			uap: '5381590794', // FIXME
+			lineItemId: '5381590794', // FIXME
+			creativeId: 'foo', // FIXME
```
[the remark this answers is not in the document]
Comment (fraszczakszymon) [reply src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts:103]: We're going to replace them once we have a way to fetch config (ADEN-10312)

File: src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+				background: platformConfig.images.default,
+			},
+			image2: {
+				element: this.resolvedBackground,
+				background: platformConfig.images.resolved,
+			},
+			adContainer: this.adContainer,
+			thumbnail: this.thumbnail,
+
+			// vastUrl: config.vast,  // FIXME
```
[the remark this answers is not in the document]
Comment (fraszczakszymon) [reply src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts:123]: Ditto ;)

File: src/ad-engine/services/template-service.ts [hunk 8]
```diff
@@ -54,6 +54,22 @@ class TemplateService {
 
 		return new this.templates[name](slot).init(params);
 	}
+
+	registerMessageListener(): void {
```
[the remark this answers is not in the document]
Comment (fraszczakszymon) [reply src/ad-engine/services/template-service.ts:58]: Good idea 👍  I'm going to replace it.

File: src/ad-engine/utils/iframe-builder.ts (hunk 2 above)
Comment (fraszczakszymon) [reply src/ad-engine/utils/iframe-builder.ts:2]: 👍

File: src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+			videoAspectRatio: 16 / 9,
+			theme: 'hivi',
+			isDarkTheme: false,
+			blockOutOfViewportPausing: true,
+			clickThroughURL: this.config.clickThroughUrl,
+			fullscreenable: true,
+		});
+	}
+
+	private getStateSetup(deviceType: string) {
```
[the remark this answers is not in the document]
Comment (fraszczakszymon) [reply src/ad-products/templates/safe-uap/safe-big-fancy-ad-proxy.ts]: Done 👍

**Pull request merged**